pkg/proxy/endpoints.go

```diff
@@ -561,8 +567,21 @@ func detectStaleConnections(oldEndpointsMap, newEndpointsMap EndpointsMap, stale
 			continue
 		}
 
-		// For udp service, if its backend changes from 0 to non-0. There may exist a conntrack entry that could blackhole traffic to the service.
-		if len(epList) > 0 && len(oldEndpointsMap[svcPortName]) == 0 {
+		epReady := 0
+		for _, ep := range epList {
+			if ep.IsReady() {
+				epReady++
```

Comment:
So... this doesn't take `ProxyTerminatingEndpoints` into account...

Consider the case where an `externalTrafficPolicy: Local` service has a single Serving-Terminating endpoint. Connections come in to that endpoint's node and are accepted and processed by the terminating pod. Then a new endpoint starts up and becomes Ready. Given the code here, that would be interpreted as "the service went from 0 endpoints to non-0 endpoints", and so the node with the Serving-Terminating endpoint would flush all conntrack entries for the service, breaking the existing connections to the Serving-Terminating pod.

(Also, this patch changes the rules for `staleServices`, but there are terminating endpoints problems with `staleEndpoints` too; we used to delete conntrack entries to endpoints as soon as the endpoint become non-ready, but now we don't delete them until the pod is fully deleted...)

Reply:
> So... this doesn't take `ProxyTerminatingEndpoints` into account...

this is a regression that needs to be backported  and ProxyTerminatingEndpoints is an alpha feature. Also, after the analysis you did in your related PR, I don't think that is easy to solve both problems easily at the same time :sweat_smile: 

>  breaking the existing connections to the Serving-Terminating pod.

it is UDP, in the sense that is not breaking the connection per se, the new packet will create a new entry looking at the iptables rules, that should still exist ... is less performant because but not a big deal

> (Also, this patch changes the rules for staleServices, but there are terminating endpoints problems with staleEndpoints too; we used to delete conntrack entries to endpoints as soon as the endpoint become non-ready, but now we don't delete them until the pod is fully deleted...)

that is fixed by the `Equal` change to take into consideration Ready https://github.com/kubernetes/kubernetes/pull/106163/files#r743336980

Reply:
> it is UDP, in the sense that is not breaking the connection per se, since it is connectionless

UDP is connectionless at L4, but not necessarily at L7. That's the main reason UDP conntrack records exist. Eg, anything using Datagram TLS (like QUIC / HTTP/3) won't survive being switched to a different endpoint mid-communication, because the new endpoint won't have the encryption key it needs.

Reply:
:eyes: I can't argue against that, but seems we are going to have some fun soon :smile: